### PR TITLE
[FEAT] LRU Caching of user files

### PR DESCRIPTION
**Pull request by dyst5422**
Caches as long as the user file does not have any type errors. Current default TTL is 30 minutes

Addresses #2

Slight API change required for implementing the caching without singleton behavior.

Drop in replacement of the call see https://github.com/NASA-AMMOS/aerie-ts-user-code-runner/pull/3/commits/705e39bec8b450a4a3fe720dbd675517fd332d01#diff-2d1b7b303360418c0f71b37dd57336aeddc1584a2668d042a44f923e70ae48b4R22-R24

The switch to a class was motivated by the desire to attach the LRU cache to an instance (and not dangling in the module).

### CODE REVIEW

**Review by pcrosemurgy — APPROVED** [review]

File: src/UserCodeRunner.ts
Comment (pcrosemurgy) [inline src/UserCodeRunner.ts:66]: Looks like this file is using more spaces per tab than the other sources.

File: src/UserCodeRunner.ts
Comment (pcrosemurgy) [inline src/UserCodeRunner.ts:154]: Likely doesn't matter but on our `feature/aerie-1741-ts-activity-instances` aerie branch we're using `ES2021`. Should that be changed to `ES2022`?

Comment (goetzrrGit) [reply src/UserCodeRunner.ts:154]: We can't target ES2022 because it isn't available yet? I tried when I created the commanding server but npm yelled at me.

**Review by goetzrrGit — COMMENTED** [review]
Spent the morning looking over the `UserCodeRunner.ts` I would approve it but we should get someone else on the team to look it over.

**Review by camargo — APPROVED** [review]

File: src/UserCodeRunner.ts
Comment (mattdailis) [reply src/UserCodeRunner.ts:66]: ☝️ this makes it quite difficult to read the diff in this file

Comment (mattdailis) [reply src/UserCodeRunner.ts:66]: NVM, Github to the rescue:

Screen Shot 2022-03-21 at 10 05 52 AM

File: src/UserCodeRunner.ts
Comment (mattdailis) [inline src/UserCodeRunner.ts:204]: Is the user file cache cleared on every run, or does it hang around as long as the service is running?

I'm wondering about using the user code hash as a key - the same expansion logic can be run against multiple mission models, right? And it's possible that it would successfully compile for one of them, and have compilation errors for another. If the cache is re-used across expansion runs, my understanding of this code is that the same expansion logic will be compiled once, and then never again, even if it's run against a different mission model. If that's the case, we will miss potential compilation errors against the subsequent mission models.

**Review by mattdailis — APPROVED** [review]
Approving for merge to unblock the command expansion effort, but in the near future I would like to continue the discussion about what to use as a key for the cache.

File: test/UserCodeRunner.spec.ts
Comment (mattdailis) [inline test/UserCodeRunner.spec.ts:232]: FYI the tests fail here with a "no such file" error. Did we forget to commit something, or should this file be generated by the test?

```
ENOENT: no such file or directory, open '/Users/dailis/projects/AERIE/aerie-ts-user-code-runner/test/inputs/command-types.ts'
```

Screen Shot 2022-03-21 at 10 52 51 AM

File: src/UserCodeRunner.ts
Comment (mattdailis) [reply src/UserCodeRunner.ts:204]: A related edge case that comes to mind is that a user can submit the same expansion logic for multiple activity types. If the first one typechecks successfully, none of the others will be typechecked at all, even though they would probably fail, since they use different activity types.

I think what I'm arriving at is that the `user_file_cache` is meant to be a memoization of the `preProcess` function, so its key should include all of the arguments to `preProcess` (notably the additional source files)

**Comment by pcrosemurgy** [comment]
~~Seem to be a fair number of open questions, probably best to wait for Dylan before taking any action. Luckily this is an optimization and not blocking other issues :)~~

The title and description of this PR may need updating. This PR fixes some fairly critical parts of the VM code runner. Other PRs will be blocked on this.

File: src/UserCodeRunner.ts
Comment (goetzrrGit) [reply src/UserCodeRunner.ts:204]: Talking with Dylan last week about it, it seems like everything is cached as long as the service is running. When reading through the code, the LRU system will remove very old runs if the queue is full or if a run is more than  ~21 days old.

I hope the below helps explain things better, This is the most confusing part of the system.

In the expansion process, the user will need to define an [expansion_set](https://github.com/NASA-AMMOS/aerie/wiki/Create-an-Expansion-Set)

When the user does an `expansion run` you tell the run which `expansion_set` to use. Let's say the logic is wrong and you get an error. Then the user will have to go back and either redefine an expansion_set that will work or go back further and update the logic. 

We still don't have a versioning system for authored expansion logic so we are basically saying each logic is the latest version so use that for now in the run.

File: test/UserCodeRunner.spec.ts
Comment (goetzrrGit) [reply test/UserCodeRunner.spec.ts:232]: You probably have to get a command library from the command expansion server. 

https://github.com/NASA-AMMOS/aerie/wiki/Upload-and-Generate-Command_Library-File

Look at our documentation finally being used :) It grows up so fast.

Comment (goetzrrGit) [reply test/UserCodeRunner.spec.ts:232]: And I just realized no one has the bananation_cdict.xml  file to upload at this moment.

Comment (pcrosemurgy) [reply test/UserCodeRunner.spec.ts:232]: Should we maybe check it into VCS under `examples/banananation/`?

Comment (goetzrrGit) [reply test/UserCodeRunner.spec.ts:232]: Yeah, It is not very complete though as it only has `BakeBananaBread` commands. I will have to update it later when I add more commands

File: src/UserCodeRunner.ts
Comment (dyst5422) [reply src/UserCodeRunner.ts:204]: I think the cache is limited to 30 minutes based on the 1000 * 60 * 30 on line 71. @mattdailis you reminded me though that I wanted to hash based not only on user code, but all included files and input/output values. I think I forgot about this when implementing. Should be as simple as changing line 202 to something like 
```
const userCodeHash = UserCodeRunner.hash(`${userCode}:${outputType}:${inputTypes.join(':')}${additionalSourceFiles.map(f => `:${f.sourceText()}`).join('')}`);
```

**Pull request merged**
